quic/s2n-quic-core/src/stream/state.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::ensure;
+use core::fmt;
+
+pub type Result<T> = core::result::Result<(), Error<T>>;
+
+macro_rules! transition {
+    ($state:ident, $valid:pat => $target:expr) => {{
+        ensure!(*$state != $target, Err(Error::NoOp { current: $target }));
+        ensure!(
+            matches!($state, $valid),
+            Err(Error::InvalidTransition {
+                current: $state.clone(),
+                target: $target
+            })
+        );
+        #[cfg(feature = "tracing")]
+        {
+            tracing::debug!(prev = ?$state, next = ?$target);
+        }
+        *$state = $target;
+        Ok(())
+    }};
+}
+
+macro_rules! is {
+    ($($state:ident)|+, $function:ident) => {
+        #[inline]
+        pub fn $function(&self) -> bool {
+            matches!(self, $(Self::$state)|*)
+        }
+    };
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error<T> {
+    NoOp { current: T },
+    InvalidTransition { current: T, target: T },
+}
```

Comment:
Ah, so the state transitions for streams would be influenced by the peer, so you'd need to return an error and close the connection (versus just panicking in a debug assertion). This is different than in BBR for example where if a an invalid state transition it was a software bug and we can just panic. So maybe not directly useable by BBR as is, but maybe somewhere else?

Reply:
Yeah most likely could be used elsewhere, including BBR where we just have a variant that does `debug_assertion` instead of return an error.

One nice pattern about returning an error though is you don't have to check the state before transitioning. So something like

```rust
fn on_packet_sent(&mut self, packet: &[u8], is_fin: bool) {
    // ignore if we're already in this state, just make sure we transition from Ready to Send
    let _ = self.state.on_send();

    if is_fin && self.state.on_send_fin().is_ok() {
        println!("first transmission of fin bit");
    }
}
```